src/main/java/uk/nhs/hee/tis/trainee/forms/service/PdfPublisherService.java

```diff
@@ -0,0 +1,175 @@
+    this.snsTemplate = snsTemplate;
+    this.publishTopic = publishTopic;
+
+    XRLog.setLoggerImpl(new Slf4jLogger());
+  }
+
+  /**
+   * Publish a Conditions of Joining PDF, the PDF will be uploaded to a file store and a
+   * notification sent to inform downstream services where to find it.
+   *
+   * @param signedEvent The COJ signed event to publish a PDF for.
+   * @throws IOException If a valid PDF could not be created.
+   */
+  public void publishConditionsOfJoining(ConditionsOfJoiningSignedEvent signedEvent)
+      throws IOException {
+    GoldGuideVersion version = signedEvent.conditionsOfJoining().version();
+    String traineeId = signedEvent.traineeId();
+    String programmeMembershipId = signedEvent.programmeMembershipId().toString();
+
+    log.info("Publishing a {} Conditions of Joining for trainee '{}' and programme membership '{}'",
+        version, traineeId, programmeMembershipId);
+
+    TemplateSpec templateSpec = version.getConditionsOfJoiningTemplate();
+    byte[] pdf = generatePdf(templateSpec, Map.of("event", signedEvent));
+
+    PublishedPdf pdfRef = upload(traineeId, FORM_TYPE_COJ, programmeMembershipId, pdf);
+
+    ConditionsOfJoiningPublishedEvent publishEvent = new ConditionsOfJoiningPublishedEvent(
+        signedEvent, pdfRef);
+    publish(FORM_TYPE_COJ, programmeMembershipId, publishEvent);
+  }
+
+  /**
+   * Generated a PDF version of the form.
+   *
+   * @param templateSpec      The template spec to use.
+   * @param templateVariables The variables to insert in to the template.
+   * @return The generated PDF as an array of bytes.
+   * @throws IOException If the renderer could not build a valid PDF.
+   */
+  private byte[] generatePdf(TemplateSpec templateSpec, Map<String, Object> templateVariables)
+      throws IOException {
+    log.info("Generating a PDF using template '{}'.", templateSpec.getTemplate());
+    String body = templateEngine.process(templateSpec, new Context(null, templateVariables));
+    Document parsedBody = Jsoup.parse(body);
+
+    ByteArrayOutputStream os = new ByteArrayOutputStream();
+    new PdfRendererBuilder()
+        .toStream(os)
+        .withW3cDocument(W3CDom.convert(parsedBody), "classpath:/")
+        .run();
+
+    return os.toByteArray();
+  }
+
+  /**
+   * Upload the form PDF to S3.
+   *
+   * @param traineeId The trainee ID the form belongs to.
+   * @param formType  The type of form being uploaded.
+   * @param filename  The filename without extension e.g. PM ID.
+   * @param pdf       The bytes of the PDF file.
+   * @return An object referencing the published PDF.
+   */
+  private PublishedPdf upload(String traineeId, String formType, String filename, byte[] pdf) {
+    log.info("Uploading generated {} for trainee '{}': {}.pdf", formType, filename, traineeId);
+    String key = String.format("%s/forms/%s/%s.pdf", traineeId, formType.toLowerCase(), filename);
+
+    s3Template.upload(publishBucket, key, new ByteArrayInputStream(pdf));
```

Comment:
Are we ok doing this synchronously? That would match how we upload the form json, but these might be slightly larger files.

Reply:
A hadn't even considered that the S3Template wasn't using an async client under the hood 🤦🏻 
While I'm not sure there's a massive amount of benefit right now, we'll slowly be adding PDF uploads from Form Rs and LTFT so I'll just refactor it now.